File: ruby/lesson 8/bfs_graph.rb
def traverse(graph)
    visited = Array.new(graph.size)
    (0..graph.size - 1).each { |v| bfs(graph, v, visited) unless visited[v] }
end

def bfs(graph, start, visited)
    queue = Queue.new

    visited[start] = true
    queue.push(start)
    puts(start)

    while !queue.empty?
        v = queue.shift()        
        graph.get(v).each do |w|
            next if visited[w]
            visited[w] = true
            queue.push(w)
            puts(w)
        end
    end 
end

File: javascript/lesson 8/bfsGraph.js
function traverse(graph) {
    const visited = Array(graph.size());
    for (let v = 0; v < graph.size(); v++) {
        if (!visited[v]) bfs(graph, v, visited);
    }
}

function bfs(graph, start, visited) {
    const queue = [];
    
    visited[start] = true;
    queue.push(start);    
    console.log(start);
    
    while (queue.length > 0) {
        const v = queue.shift();
        for (let w of graph.get(v)) {
            if (!visited[w]) {
                visited[w] = true;
                queue.push(w);
                console.log(w);
            }
        }
    }
}
